SONARXML-221 : S5604 Should not raise on items containing tools:node="remove"

rombirli opened the pull request.

rombirli committed:
Draft of SONARXML-221 solution

tomasz-tylenda-sonarsource commented on sonar-xml-plugin/src/main/java/org/sonar/plugins/xml/checks/security/android/AndroidPermissionsCheck.java:
```diff
@@ -111,12 +114,26 @@ public class AndroidPermissionsCheck extends AbstractAndroidManifestCheck {
     evaluateAsList(xPathExpression, file.getDocument()).stream()
-      .filter(node -> DANGEROUS_PERMISSIONS.contains(node.getNodeValue()))
-      .forEach(node -> reportIssue(node, String.format(MESSAGE, simpleName(node.getNodeValue()))));
+      .filter(node -> !hasToolsNodeRemove(node))
```
It might be a little more readable to use one `filter` with bodies merged with `&&`.

dorian-burihabwa-sonarsource commented on sonar-xml-plugin/src/main/java/org/sonar/plugins/xml/checks/security/android/AndroidPermissionsCheck.java:
```diff
@@ -111,12 +114,26 @@ public class AndroidPermissionsCheck extends AbstractAndroidManifestCheck {
     return fullyQualifiedName.substring(fullyQualifiedName.lastIndexOf('.') + 1);
   }
 
+  private static Node findPermissionNode(Node node) {
```
Is this method returning a node or an attribute? It looks like we pass a `<uses-permission> node as argument and retrieve an attribute as a result?
```suggestion
  private static Node findPermissionNameAttribute(Node node) {
```

rombirli replied:
The attribute is a node of the AST but not an XML node effectively so your name proposal makes more sense. Applied

dorian-burihabwa-sonarsource reviewed (changes requested):
This solution looks pretty solid. I think there are a couple of things we could do to get to the extra mile:
1. Cover `removeAll` like we cover `remove`
2. Rename the new test and new utility methods in the check

rombirli committed:
Refactoring : apply most change requests

rombirli committed:
Add tools:node="removeAll" support - no caching

rombirli committed:
Caching for tools:node="removeAll"

rombirli committed:
Refactoring : rename tests

rombirli commented on sonar-xml-plugin/src/main/java/org/sonar/plugins/xml/checks/security/android/AndroidPermissionsCheck.java:
```diff
@@ -111,12 +114,26 @@ public class AndroidPermissionsCheck extends AbstractAndroidManifestCheck {
+
+  private static boolean hasToolsNodeRemove(Node node) {
+    Node toolsNodeAttribute = node.getAttributes().getNamedItemNS(ANDROID_MANIFEST_TOOLS, "node");
+    return Optional.ofNullable(toolsNodeAttribute).map(n -> "remove".equals(n.getNodeValue())).orElse(false);
```
Added support and tests for removeAll

dorian-burihabwa-sonarsource reviewed (changes requested):
The approach looks good but there are a couple of things we need to fix before we can merge.

rombirli commented on sonar-xml-plugin/src/main/java/org/sonar/plugins/xml/checks/security/android/AndroidPermissionsCheck.java:
```diff
@@ -110,13 +119,55 @@ public class AndroidPermissionsCheck extends AbstractAndroidManifestCheck {
+    Node permissionsAttribute = findPermissionAttribute(node);
+    String permissionValue = permissionsAttribute.getNodeValue();
+    if (!hasToolsNodeRemove(node)
+      && !hasParentToolsNodeRemoveAll(node)
```
If I understand correctly your approach, a tools:node="removeAll" removes all emenent with tag:uses-permission AND same name, but I understand the [documentation](https://developer.android.com/build/manage-manifests#node_markers)  differently, for what i understood, the name is not checked, and elements with same tag in same parent are removed

rombirli committed:
Refactoring, rename tests, remove cache, remove repetitions

dorian-burihabwa-sonarsource commented on sonar-xml-plugin/src/main/java/org/sonar/plugins/xml/checks/security/android/AndroidPermissionsCheck.java:
```diff
@@ -110,13 +118,44 @@ public class AndroidPermissionsCheck extends AbstractAndroidManifestCheck {
-    evaluateAsList(xPathExpression, file.getDocument()).stream()
-      .filter(node -> DANGEROUS_PERMISSIONS.contains(node.getNodeValue()))
-      .forEach(node -> reportIssue(node, String.format(MESSAGE, simpleName(node.getNodeValue()))));
+    if (noToolsNodeRemoveAll(file.getDocument().getFirstChild())) {
```
We need to be careful here with the first element. We know we entered `scanAndroidManifest` here because the file is appropriately named `AndroidManifest.xml` thanks to this [test](https://github.com/SonarSource/sonar-xml/blob/d4db3cdd687cee86fe0c67f5dbb5343ff6b06c4a/sonar-xml-plugin/src/main/java/org/sonar/plugins/xml/checks/security/android/Utils.java#L29).
But it does not tell us that the AndroidManifest file is structurally correct. We could be looking at a random first element with a large number of child nodes and wasting time here.

I think we could fix this  one of 2 ways:
1. by replacing `file.getDocument().getFirstChild()` with the first path of the xpath query `evaluateAsList("/manifest", file.getDocument())` and check if we have an element
2.  by capturing all of the `manifest/uses-permission` under nodes before doing any test and passing that list to `noTollsNodeRemoveAll` (and maybe re-using that same list in the block under this test)

dorian-burihabwa-sonarsource replied:
Choosing one of these 2 options should also help with coverage

rombirli replied:
chose option 2 as it required fewer changes

dorian-burihabwa-sonarsource reviewed (changes requested):
Thank you for the changes, we are almost there and should be getting there soon.

rombirli committed:
Refactoring : noToolsNodeRemoveAll signature change
 - remove @nullable decorator on node argument
 - invert return type, change name accordingly
 - takes uses-permission nodes list instead of manifest node as argument

dorian-burihabwa-sonarsource reviewed (approved):
Great job on this one! The checks looks a lot more approachable with this version.

rombirli committed:
Make function hasToolsNodeRemoveAll static
